go/vt/vtgate/engine/route.go

```diff
@@ -613,7 +619,7 @@ func (route *Route) resolveSingleShard(vcursor VCursor, bindVars map[string]*que
 		return "", "", nil, err
 	}
 	ksid = ksids[0]
-	if len(ksid) == 0 {
+	if ksid == nil {
```

Comment:
There is a difference:
```
  var b []byte // b == nil, len(b) == 0
  var c = []byte{} // c != nil, len(c) == 0
```

In the case of keyspace ids, a nil []byte means that no match was found like `b`. But an empty value like `c` means that a keyspace id of 0-length was  found.

Right now, I think we consider empty keyspace ids to be valid, they go to shard 0. But I agree this is confusing, and we can look at disallowing this. This will allow us to treat empty and nil to be the same.